File: netlogo-gui/src/main/window/DummyMonitorWidget.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.window;

import org.nlogo.api.Editable;
import org.nlogo.core.I18N;
import org.nlogo.api.Property;

import java.util.List;

public strictfp class DummyMonitorWidget
    extends SingleErrorWidget
    implements
    Editable {

  private static final int LEFT_MARGIN = 5;
  private static final int RIGHT_MARGIN = 6;
  private static final int BOTTOM_MARGIN = 6;

  public DummyMonitorWidget() {
    super();
    setOpaque(true);
    setBackground(InterfaceColors.MONITOR_BACKGROUND);
    setBorder(widgetBorder());
    org.nlogo.awt.Fonts.adjustDefaultFont(this);
  }

  private String name = "";

  public void name(String name) {
    this.name = name;
  }

  public String name() {
    return name;
  }

  @Override
  public String displayName() {
    return name;
  }

  @Override

  public String classDisplayName() {
    return I18N.guiJ().get("tabs.run.widgets.monitor");
  }

  public List<Property> propertySet() {
    return Properties.dummyMonitor();
  }

  private static final int MIN_WIDTH = 50;
  // private static final int MIN_HEIGHT = 35 ; // no longer used
  private static final int MAX_HEIGHT = 49;

  @Override
  public java.awt.Dimension getMinimumSize() {
    return new java.awt.Dimension(MIN_WIDTH, MAX_HEIGHT);
  }

  @Override
  public java.awt.Dimension getMaximumSize() {
    return new java.awt.Dimension(10000, MAX_HEIGHT);
  }

  @Override
  public java.awt.Dimension getPreferredSize(java.awt.Font font) {
    java.awt.Dimension size = getMinimumSize();
    int pad = 12;
    java.awt.FontMetrics fontMetrics = getFontMetrics(font);
    size.width = StrictMath.max(size.width, fontMetrics.stringWidth(displayName()) + pad);
    size.height = StrictMath.max(size.height, fontMetrics.getMaxDescent() + fontMetrics.getMaxAscent() + pad);
    return size;
  }

  @Override
  public void paintComponent(java.awt.Graphics g) {
    super.paintComponent(g); // paint background
    java.awt.FontMetrics fm = g.getFontMetrics();
    int labelHeight = fm.getMaxDescent() + fm.getMaxAscent();
    java.awt.Dimension d = getSize();
    g.setColor(getForeground());
    String displayName = displayName();
    int boxHeight = (int) StrictMath.ceil(labelHeight * 1.4);
    g.drawString(displayName, LEFT_MARGIN,
        d.height - BOTTOM_MARGIN - boxHeight - fm.getMaxDescent() - 1);
    g.setColor(java.awt.Color.WHITE);
    g.fillRect(LEFT_MARGIN, d.height - BOTTOM_MARGIN - boxHeight,
        d.width - LEFT_MARGIN - RIGHT_MARGIN - 1, boxHeight);
    g.setColor(java.awt.Color.BLACK);
  }

  private int decimalPlaces = 3;

  public int decimalPlaces() {
    return decimalPlaces;
  }

  public void decimalPlaces(int decimalPlaces) {
    if (decimalPlaces != this.decimalPlaces) {
      this.decimalPlaces = decimalPlaces;
    }
  }

  @Override
  public String save() {
    StringBuilder s = new StringBuilder();
    s.append("MONITOR\n");
    s.append(getBoundsString());
    if ((null != name()) && (!name().trim().equals(""))) {
      s.append(name() + "\n");
    } else {
      s.append("NIL\n");
    }
    s.append("NIL\n");

    s.append(decimalPlaces + "\n");
    s.append("1\n");  // for compatability

    return s.toString();
  }

  @Override
  public Object load(String[] strings, Widget.LoadHelper helper) {
    String displayName = strings[5];

    if (displayName.equals("NIL")) {
      name("");
    } else {
      name(displayName);
    }
    if (strings.length > 7) {
      decimalPlaces = Integer.parseInt(strings[7]);
    }
    int x1 = Integer.parseInt(strings[1]);
    int y1 = Integer.parseInt(strings[2]);
    int x2 = Integer.parseInt(strings[3]);
    int y2 = Integer.parseInt(strings[4]);
    setSize(x2 - x1, y2 - y1);
    return this;
  }
}

File: netlogo-gui/src/main/window/DummyMonitorWidget.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.window

import java.awt.{ Dimension, Insets }
import javax.swing.{ BoxLayout, JLabel, JPanel }

import org.nlogo.core.{ I18N, Monitor => CoreMonitor, Widget => CoreWidget }
import org.nlogo.swing.{ BoxAlign, BoxRow, RoundedBorderPanel, Utils }
import org.nlogo.theme.{ InterfaceColors, ThemeSync }

class DummyMonitorWidget extends SingleErrorWidget with MonitorWidget.ToMonitorModel with Editable {
  private var _name: String = ""
  private var _decimalPlaces = 3
  private var _units: String = ""

  private val nameLabel = new JLabel(I18N.gui.get("edit.monitor.previewName"))

  private val valuePanel = new JPanel with RoundedBorderPanel with ThemeSync {
    setDiameter(6)

    override def syncTheme(): Unit = {
      setBackgroundColor(InterfaceColors.displayAreaBackground())
      setBorderColor(InterfaceColors.monitorBorder())
    }
  }

  setLayout(new BoxLayout(this, BoxLayout.Y_AXIS))
  setBorder(new AdaptableBorder(new Insets(3, 6, 6, 6), new Insets(6, 8, 8, 8)))

  add(new BoxRow(nameLabel, BoxAlign.Start))
  add(new AdaptableVerticalStrut(0, 6))
  add(valuePanel)

  def innerSource = ""
  def fontSize = 11

  def name: String = _name

  def setDisplayName(name: String): Unit = {
    val suffix = if (_units.isEmpty) "" else (" " + _units)
    _name = name + suffix
    displayName = name

    if (_name.trim.isEmpty) {
      nameLabel.setText(I18N.gui.get("edit.monitor.previewName"))
    } else {
      nameLabel.setText(_name)
    }
  }

  def units: String = _units
  def setUnits(value: String): Unit = {
    _units = value
    revalidate()
    repaint()
  }

  override def classDisplayName: String =
    I18N.gui.get("tabs.run.widgets.monitor")

  override def editPanel: EditPanel = new DummyMonitorEditPanel(this)

  override def getEditable: Option[Editable] = Some(this)

  override def getMinimumSize: Dimension = {
    if (_oldSize) {
      new Dimension(Utils.zoom(50), (fontSize * 4) + Utils.zoomClamped(1))
    } else {
      new Dimension(Utils.zoom(100), Utils.zoom(60))
    }
  }

  override def getPreferredSize: Dimension = {
    if (_oldSize) {
      new Dimension(Utils.zoom(100), getMinimumSize.height)
    } else {
      new Dimension(Utils.zoom(100), Utils.zoom(60))
    }
  }

  override def syncTheme(): Unit = {
    setBackgroundColor(InterfaceColors.monitorBackground())

    nameLabel.setForeground(InterfaceColors.widgetText())

    valuePanel.syncTheme()
  }

  def decimalPlaces: Int = _decimalPlaces

  def setDecimalPlaces(decimalPlaces: Int): Unit = {
    if (decimalPlaces != _decimalPlaces)
      _decimalPlaces = decimalPlaces
  }

  override def load(model: CoreWidget): Unit = {
    model match {
      case monitor: CoreMonitor =>
        setUnits(monitor.units.getOrElse(""))
        setDisplayName(monitor.display.optionToPotentiallyEmptyString)
        setDecimalPlaces(monitor.precision)
        oldSize(monitor.oldSize)
        setSize(monitor.width, monitor.height)

      case _ =>
    }
  }
}
